File: LeetCode_Java/interval/NonOverlappingIntervals.java
package interval;

import java.util.Arrays;

public class NonOverlappingIntervals {
    public int eraseOverlapIntervals(int[][] intervals) {

        int n = intervals.length;
        if (n == 1) {
            return 0;
        }

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[1], b[1]));

        int count = 0;
        int end = intervals[0][1];

        for (int i = 1; i < n; i++) {
            if (intervals[i][0] < end) {
                count++;
            } else {
                end = intervals[i][1];
            }
        }

        return count;
    }

    public static void main(String[] args) {
        NonOverlappingIntervals solution = new NonOverlappingIntervals();

        int[][] intervals1 = {{1, 2}, {2, 3}, {3, 4}, {1, 3}};
        System.out.println(solution.eraseOverlapIntervals(intervals1)); // Output: 1

        int[][] intervals2 = {{1, 2}, {1, 2}, {1, 2}};
        System.out.println(solution.eraseOverlapIntervals(intervals2)); // Output: 2

        int[][] intervals3 = {{1, 2}, {2, 3}};
        System.out.println(solution.eraseOverlapIntervals(intervals3)); // Output: 0
    }
}

File: LeetCode_Python/intervals/NonOverlappingIntervals.py
class NonOverlappingIntervals:
    def eraseOverlapIntervals(self, intervals: list[list[int]]) -> int:
        intervals.sort(key=lambda x: x[1])

        res = 0
        end = intervals[0][1]

        for i in range(1, len(intervals)):
            if intervals[i][0] < end:
                res += 1
            else:
                end = intervals[i][1]

        return res


if __name__ == "__main__":
    solution = NonOverlappingIntervals()

    print("Expected Output: 1")
    print(
        "Actual Output: ",
        solution.eraseOverlapIntervals([[1, 2], [2, 3], [3, 4], [1, 3]]),
    )

    print("Expected Output: 2")
    print("Actual Output: ", solution.eraseOverlapIntervals([[1, 2], [1, 2], [1, 2]]))

    print("Expected Output: 0")
    print("Actual Output: ", solution.eraseOverlapIntervals([[1, 2], [2, 3]]))
